core/src/main/java/tech/pegasys/web3signer/core/service/http/handlers/keymanager/imports/ImportKeystoresHandler.java

```diff
@@ -0,0 +1,250 @@
+import tech.pegasys.web3signer.core.multikey.metadata.SignerOrigin;
+import tech.pegasys.web3signer.core.signing.ArtifactSignerProvider;
+import tech.pegasys.web3signer.core.signing.BlsArtifactSigner;
+import tech.pegasys.web3signer.core.signing.KeyType;
+import tech.pegasys.web3signer.slashingprotection.SlashingProtection;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
+import io.vertx.core.Handler;
+import io.vertx.core.json.JsonObject;
+import io.vertx.ext.web.RoutingContext;
+import io.vertx.ext.web.api.RequestParameters;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.tuweni.bytes.Bytes;
+import org.apache.tuweni.bytes.Bytes32;
+
+public class ImportKeystoresHandler implements Handler<RoutingContext> {
+
+  private static final Logger LOG = LogManager.getLogger();
+  public static final int SUCCESS = 200;
+  public static final int BAD_REQUEST = 400;
+  public static final int SERVER_ERROR = 500;
+  private static final ObjectMapper YAML_OBJECT_MAPPER = new ObjectMapper(new YAMLFactory());
+
+  private final ObjectMapper objectMapper;
+  private final Path keystorePath;
+  private final Optional<SlashingProtection> slashingProtection;
+  private final ArtifactSignerProvider artifactSignerProvider;
+
+  public ImportKeystoresHandler(
+      final ObjectMapper objectMapper,
+      final Path keystorePath,
+      final Optional<SlashingProtection> slashingProtection,
+      final ArtifactSignerProvider artifactSignerProvider) {
+    this.objectMapper = objectMapper;
+    this.keystorePath = keystorePath;
+    this.slashingProtection = slashingProtection;
+    this.artifactSignerProvider = artifactSignerProvider;
+  }
+
+  @Override
+  public void handle(RoutingContext context) {
+    // API spec - https://github.com/ethereum/keymanager-APIs/tree/master/flows#import
+    final RequestParameters params = context.get("parsedParameters");
+    final ImportKeystoresRequestBody parsedBody;
+    try {
+      parsedBody = parseRequestBody(params);
+    } catch (final IllegalArgumentException | JsonProcessingException e) {
+      handleInvalidRequest(context, e);
+      return;
+    }
+
+    // check that keystores have matching passwords
+    if (parsedBody.getKeystores().size() != parsedBody.getPasswords().size()) {
+      context.fail(BAD_REQUEST);
+      return;
+    }
+
+    // no keystores passed in, nothing to do, return 200 and empty response.
```

Comment:
I went back and forth on this because according to the discussion with the validation clients teams, filtering the slashing data is a "nice to have" but not required, I think that's why the spec is contradicting itself here. I know Teku has implemented it in a way that only imports slashing data for the successfully imported pubkeys, but lighthouse doesn't and just imports it all for example.

I'm on the fence about this, since doing it the Teku way does require some fairly significant refactoring to the slashing data importer. I tried to minimize refactors in my implementation, which ends up being somewhere in the middle, doesn't import any random data (only pubkeys present in the request), but as you correctly pointed out, it's possible that some of those keys don't get imported successfully.

My hope is that we can use this approach for now and do the slashing data importer refactoring later, but let me know what you prefer.

Reply:
Thanks for the explanation. I get your point about the refactoring and agree it might be better handled separately. 

> it's possible that some of those keys don't get imported successfully

If I've understood correctly then in this scenario, you'll get slashing data (unnecessarily) imported but still receive a 200 response (with a keystore failure message) so it's not like we're performing actions on the server and responding with an error which is good.
As long as the state this leaves the DB in won't cause problems for future imports then I'm OK with this, e.g. same/new slashing data + a now valid key.

Reply:
yeah it should be fine to import more slashing data than necessary. My understanding is that this data is to make sure we don't sign the same thing twice, in which case having more data in the DB shouldn't hurt.

And yes subsequent calls with the same/new slashing data with now valid keys should just re-import that same data fine.